train/compute/python/tools/trace_link.py

```diff
@@ -0,0 +1,309 @@
+                kineto_nodes.append(tmp)
+
+    # Max call stack depth when building the tree, the deeper the more accurate but takes longer time
+    depth = 10
+
+    # Build a tree from the kineto trace
+    kineto_graph = transform_to_graph_depth(process_node, depth)
+    print("Kineto tree nodes number: ", len(kineto_graph.nodes))
+
+    # Build a tree from the execution trace
+    et_graph = transform_to_graph_depth(et_nodes[0], depth)
+    print("ET tree nodes number: ", len(et_graph.nodes))
+
+    # print("Kineto nodes info: ", list(kineto_graph.nodes(data=True)))
+    # print("ET nodes info: ", list(et_graph.nodes(data=True)))
+
+    # Create the GraphMatcher
+    GM = isomorphism.GraphMatcher(kineto_graph, et_graph)
+
+    # Duration of ET nodes
+    et_enhanced = {}
+
+    if GM.is_isomorphic():
+        mapping = GM.mapping
+        print("Graphs are isomorphic")
+        for kineto_id, et_id in mapping.items():
+            et_enhanced[et_id] = kineto_nodes_mapping[kineto_id].end - kineto_nodes_mapping[kineto_id].start
+    else:
+        print("Graphs are not isomorphic")
+
+        # # Compute the edit distance using the graph_edit_distance function with node comparison
+        # paths, cost = nx.graph_edit_distance(kineto_graph, et_graph, node_compare)
+        # print("Tree edit distance:", cost)
+
+        # The problem of finding the exact Graph Edit Distance (GED) is NP-hard so it is often slow
+        # and below is a sub-optimal approach
+
+        edit_distance_generator = nx.optimize_graph_edit_distance(kineto_graph, et_graph, node_compare)
+        cost = next(edit_distance_generator)
+
+        paths_generator = nx.optimize_edit_paths(kineto_graph, et_graph, node_compare)
+        node_edits, edge_edits, cost = next(paths_generator)
+
+        print("Sub-optimal tree edit distance:", cost)
+
+        # print("node_edits: ", node_edits)
+
+        for kineto_id, et_id in node_edits:
+            if kineto_id is not None and et_id is not None:
+                et_enhanced[et_id] = kineto_nodes_mapping[kineto_id].end - kineto_nodes_mapping[kineto_id].start
+
+    return et_enhanced
+
+
+if __name__ == "__main__":
+    import argparse
+    
+    parser = argparse.ArgumentParser(description="Link kineto trace with execution trace")
+    parser.add_argument("--et_file", type=str, required=True, help="Path to the execution trace file")
+    parser.add_argument("--kineto_file", type=str, required=True, help="Path to the kineto trace file")
+    parser.add_argument("--annotation", default='DataLoader', type=str, help="User annotation of the iteration step of execution trace")
+    parser.add_argument("--exact_match", default=False, action='store_true', help="Whether to match the traces exactly")
+
+    args = parser.parse_args()
+
+    # Analysis of execution trace
+    with open(args.et_file, 'r') as f:
+        et = ExecutionGraph(json.load(f))
+
+    nodes = et.get_nodes(clean=True)
```

Comment:
You are correct, clean is not necessary and actually we do need the dataloader ops. Luckily here we only use it to find the root node so it will not cause issues. Will change to False to avoid the confusion.